velox/dwio/parquet/thrift/ThriftStreamingTransport.h

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <thrift/transport/TVirtualTransport.h>
+#include "velox/dwio/common/BufferedInput.h"
+#include "velox/dwio/parquet/thrift/ThriftTransport.h"
+
+namespace facebook::velox::parquet::thrift {
+
+class ThriftStreamingTransport : public ThriftTransport {
+ public:
+  ThriftStreamingTransport(
+      std::shared_ptr<dwio::common::SeekableInputStream> inputStream,
+      const char* FOLLY_NULLABLE& bufferStart,
+      const char* FOLLY_NULLABLE& bufferEnd)
+      : ThriftTransport(),
+        inputStream_(inputStream),
+        bufferStart_(bufferStart),
+        bufferEnd_(bufferEnd) {
+    VELOX_CHECK_NOT_NULL(inputStream_);
+    VELOX_CHECK_NOT_NULL(bufferStart_);
+    VELOX_CHECK_NOT_NULL(bufferEnd_);
+  }
+
+  uint32_t read(uint8_t* outputBuf, uint32_t len) {
+    uint32_t remainLen = len;
+    while (remainLen > 0) {
+      uint32_t size = bufferEnd_ - bufferStart_;
+      auto copySize = std::min(size, remainLen);
+      memcpy(outputBuf, bufferStart_, copySize);
+      remainLen -= copySize;
+      offset_ += copySize;
+      bufferStart_ += copySize;
+      outputBuf += copySize;
+      if (bufferStart_ == bufferEnd_) {
+        int32_t batchSize;
+        const void* bufferPointer;
+        if (!inputStream_->Next(&bufferPointer, &batchSize)) {
+          //          The inputStream_ is fully consumed but reading does not
+          //          finish.
+          if (remainLen > 0) {
+            VELOX_FAIL("Reading Exceeds inputStream_ Capacity");
```

Comment:
We should just return the number of bytes read so far instead of throwing.

Reply:
This means that the existing input (after consuming the entire `inputStream_`) is shorter than the required `len` in `read()`. When reading a Parquet file, it might happen when the file is corrupted.

Reply:
We need to check the actual read length is equal to the input length then.  From the perspective of this class, it's not something exceptional.

Reply:
I see. The input here is `SeekableInputStream`. Normally, it contains the whole Parquet column chunk, which should be larger than the required `len`. And when `read()`, the current `SeekableInputStream` batch might not be enough and we need to call `Next()` for next batch. Also, one Parquet page cannot exist on more one column chunk.